### PR TITLE
[SM-108] Refactor Reports - Add storybook stories

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Pair programming with @coroiu.

The focus of this PR was two-fold.

1. We wanted to add storybook stories to the Reports components.
2. Investigate if we can refactor them to be more generic.

This lead to us introducing a couple of structural changes.

1. Created a `ReportsModule`, this module imports the `ReportsRoutingModule`. This properly encapsulates the logic of the reports, and they no longer exists in the `LooseComponentsModule`.
4. Introduces a `ReportsHomeComponent` which takes over some responsibilities from from the `ReportsListComponent`. This allows the `ReportListComponent` to be generic and allows for it to be re-used in other places.
5. Moves all the "page" components to `pages`. I'm not completely happy with this folder name but it seemed better than having them in the root of the folder for now.

We also made some significant changes to the `ReportCard` and `ReportList` components. Initially we ended up with the following hierarchical relationship between components. `ReportsHomeComponent (smart) -> ReportList (presentational) -> ReportCard (smart)`. This is sub-optimal since the logic is split up into multiple places and we can't consider `ReportList` to be a strict presentational component since it directly uses a smart sub component.

In an effort to solve this we made a couple of refactors.

1. Remove the smart logic of the `ReportCard` and made it a "dumb" presentational component.
	* Changed the inputs from a single `type` which used to be looked up in a lookup table to have individual inputs for each value.
	* Introduced `ReportVariant` which denotes which badge the card should use.
	* Extracted icon logic to a separate component which allowed us to get rid of the SVG sanitation.
	* These changes had the benefits of drastically simplifying the mocks we needed in our stories.
2. Updated the `ReportList` component to accept a list of reports.
	* The `ReportList` is a strictly presentational component and is responsible for styling the grid, and populating it with the array of reports.
	* The list owns it's input model which is `ReportEntry`.
3. The actual report defintions moved to `reports.ts`.
6. The `ReportsHomePage` is responsible for building the `reports` list which serves as the input for the `ReportList` component. And it's also the only place where we need a smart component, namely to figure if the cards should display a "premium" or "upgrade" badge.
7. The premium check got moved to a route guard.

### Future Tasks

We should refactor the page components to behave better and not need to be extended for organization reports.

## Before you submit

```
- [ ] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

File: apps/web/src/app/organizations/tools/exposed-passwords-report.component.ts [hunk 1]
```diff
@@ -11,11 +11,11 @@ import { StateService } from "@bitwarden/common/abstractions/state.service";
 import { Cipher } from "@bitwarden/common/models/domain/cipher";
 import { CipherView } from "@bitwarden/common/models/view/cipherView";
 
-import { ExposedPasswordsReportComponent as BaseExposedPasswordsReportComponent } from "../../reports/exposed-passwords-report.component";
+import { ExposedPasswordsReportComponent as BaseExposedPasswordsReportComponent } from "../../reports/pages/exposed-passwords-report.component";
 
 @Component({
   selector: "app-org-exposed-passwords-report",
-  templateUrl: "../../reports/exposed-passwords-report.component.html",
+  templateUrl: "../../reports/pages/exposed-passwords-report.component.html",
```
Comment (addisonbeck) [inline apps/web/src/app/organizations/tools/exposed-passwords-report.component.ts:18]: I don't really like this subdirectory name `pages/`. `components/` makes more sense in an angular application, as a lot of the time we might want to break up code that isn't a full page. Can we standardize and call this `components/`?

File: apps/web/src/app/reports/report-list/report-list.component.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+import { ReportVariant } from "../report-card/report-card.component";
+
+export type ReportEntry = {
+  title: string;
+  description: string;
+  route: string;
+  icon: string;
+  variant: ReportVariant;
+};
```
Comment (addisonbeck) [inline apps/web/src/app/reports/report-list/report-list.component.ts:11]: Can we move this to a separate file?

File: apps/web/src/app/reports/reports.module.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    ReportCardComponent,
+    ReportListComponent,
+    ReportsLayoutComponent,
+    ReportsHomeComponent,
+    ReusedPasswordsReportComponent,
+    UnsecuredWebsitesReportComponent,
+    WeakPasswordsReportComponent,
+    WeakPasswordsReportComponent,
+  ],
+  exports: [],
```
Comment (addisonbeck) [inline apps/web/src/app/reports/reports.module.ts:33]: I think you can remove this array if it isn't used
```suggestion
```

File: apps/web/src/app/reports/reports.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import { ReportEntry } from "./report-list/report-list.component";
```
Comment (addisonbeck) [inline apps/web/src/app/reports/reports.ts:1]: Can we move `ReportEntry` to its own file?

Comment (addisonbeck) [inline apps/web/src/app/reports/reports.ts:1]: Lets move `ReportEntry` to its own file if it is going to be using outside of `ReportListComponent`

**Review by addisonbeck — COMMENTED** [review]
Great work you two! I have a few questions.

File: apps/web/src/app/organizations/tools/exposed-passwords-report.component.ts (hunk 1 above)
Comment (Hinton) [reply apps/web/src/app/organizations/tools/exposed-passwords-report.component.ts:18]: I think components would be misleading in this cases. This is a subset of components that represent entry level. We have regular components outside this folder.

File: apps/web/tsconfig.json [hunk 5]
```diff
@@ -3,6 +3,7 @@
   "compilerOptions": {
     "baseUrl": ".",
     "module": "ES2020",
+    "resolveJsonModule": true,
```
[the remark this answers is not in the document]
Comment (Hinton) [reply apps/web/tsconfig.json:6]: Allow importing json files. It's needed for the preloaded English i18n

File: apps/web/src/app/reports/report-list/report-list.component.ts (hunk 2 above)
Comment (Hinton) [reply apps/web/src/app/reports/report-list/report-list.component.ts:11]: Done.

**Review by addisonbeck — APPROVED** [review]

**Pull request merged**